src/read.c

```diff
@@ -4342,41 +4351,49 @@ static avifResult avifDecoderDataFindToneMappedImageItem(const avifDecoderData *
     return AVIF_RESULT_OK;
 }
 
-// Finds a 'tmap' (tone mapped image item) box associated with the given 'colorItem',
+// Finds a 'tmap' (tone mapped image item) box associated with the given colorItem at index 'colorItemIndex',
 // then finds the associated gain map image.
-// If found, fills 'toneMappedImageItem', 'gainMapItem' and  'gainMapCodecType'.
-// Otherwise, sets 'toneMappedImageItem' and 'gainMapItem' to NULL.
+// If found, fills 'toneMappedImageItem', 'gainMapItemIndex' and  'gainMapCodecType'.
+// Otherwise, sets 'toneMappedImageItem' to NULL and 'gainMapItemIndex' to -1.
 // Returns AVIF_RESULT_OK if no errors were encountered (whether or not a gain map was found).
 // Assumes that there is a single tmap item, and not, e.g., a grid of tmap items.
 static avifResult avifDecoderFindGainMapItem(const avifDecoder * decoder,
-                                             const avifDecoderItem * colorItem,
+                                             int colorItemIndex,
                                              avifDecoderItem ** toneMappedImageItem,
-                                             avifDecoderItem ** gainMapItem,
+                                             int * gainMapItemIndex,
                                              avifCodecType * gainMapCodecType)
 {
     *toneMappedImageItem = NULL;
-    *gainMapItem = NULL;
+    *gainMapItemIndex = -1;
     *gainMapCodecType = AVIF_CODEC_TYPE_UNKNOWN;
 
     avifDecoderData * data = decoder->data;
 
     uint32_t gainMapItemID;
     avifDecoderItem * toneMappedImageItemTmp;
-    AVIF_CHECKRES(avifDecoderDataFindToneMappedImageItem(data, colorItem, &toneMappedImageItemTmp, &gainMapItemID));
+    AVIF_CHECKRES(
+        avifDecoderDataFindToneMappedImageItem(data, &data->meta->items.item[colorItemIndex], &toneMappedImageItemTmp, &gainMapItemID));
     if (!toneMappedImageItemTmp) {
         return AVIF_RESULT_OK;
     }
 
     assert(gainMapItemID != 0);
-    avifDecoderItem * gainMapItemTmp;
-    AVIF_CHECKRES(avifMetaFindOrCreateItem(data->meta, gainMapItemID, &gainMapItemTmp));
-    if (avifDecoderItemShouldBeSkipped(gainMapItemTmp)) {
+    avifDecoderItem * gainMapItem;
```

Comment:
Nit: I wonder if the original code uses a `gainMapItemTmp` local variable so that the output parameter `*gainMapItem` is only assigned a value when the function returns AVIF_RESULT_OK. If so, the new code should do the same with the `*gainMapItemIndex` output parameter.

Reply:
The answer to my question is Yes. Please see the comment at line 4429:

```
    // Only set the output parameters after everything has been validated.
    *toneMappedImageItem = toneMappedImageItemTmp;
    return AVIF_RESULT_OK;
```

So `*gainMapItemIndex` should receive the same treatment as `*gainMapItem` in the original code.

Reply:
Done.